From: Dan Williams <dan.j.williams@intel.com>
Date: Wed, 6 Jan 2016 12:03:42 -0800
Subject: pmem, dax: disable dax in the presence of bad blocks
Git-commit: 57f7f317abdd07954cb116280c88d18378afb33e
Patch-mainline: v4.5-rc1
References: FATE#319965, bsc#964944

Longer term teach dax to punch "error" holes in mapping requests and
deliver SIGBUS to applications that consume a bad pmem page.  For now,
simply disable the dax performance optimization in the presence of known
errors.

Signed-off-by: Dan Williams <dan.j.williams@intel.com>
Acked-by: Johannes Thumshirn <jthumshirn@suse.com>
---
 block/ioctl.c         |   10 ++++++++++
 drivers/nvdimm/pmem.c |    2 ++
 2 files changed, 12 insertions(+)

--- a/block/ioctl.c
+++ b/block/ioctl.c
@@ -4,6 +4,7 @@
 #include <linux/gfp.h>
 #include <linux/blkpg.h>
 #include <linux/hdreg.h>
+#include <linux/badblocks.h>
 #include <linux/backing-dev.h>
 #include <linux/fs.h>
 #include <linux/blktrace_api.h>
@@ -422,6 +423,15 @@ bool blkdev_dax_capable(struct block_dev
 			|| (bdev->bd_part->nr_sects % (PAGE_SIZE / 512)))
 		return false;
 
+	/*
+	 * If the device has known bad blocks, force all I/O through the
+	 * driver / page cache.
+	 *
+	 * TODO: support finer grained dax error handling
+	 */
+	if (disk->bb && disk->bb->count)
+		return false;
+
 	return true;
 }
 
--- a/drivers/nvdimm/pmem.c
+++ b/drivers/nvdimm/pmem.c
@@ -295,6 +295,8 @@ static int pmem_attach_disk(struct devic
 	q->queuedata = pmem;
 
 	disk = alloc_disk_node(0, nid);
+
+	disk->bb = &pmem->bb;
 	if (!disk)
 		return -ENOMEM;
 
